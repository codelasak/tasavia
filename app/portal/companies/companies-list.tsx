'use client'

import { useEffect, useState } from 'react'
import { Button } from '@/components/ui/button'
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from '@/components/ui/card'
import { Input } from '@/components/ui/input'
import { Search, Edit, Trash2, Plus } from 'lucide-react'
import { supabase } from '@/lib/supabase/client'
import { Database } from '@/lib/supabase/database.types'
import { toast } from 'sonner'
import { CompanyDialog } from '@/components/companies/CompanyDialog'

type ExternalCompany = Database['public']['Tables']['companies']['Row'] & {
  company_contacts: Database['public']['Tables']['company_contacts']['Row'][]
  company_addresses: Database['public']['Tables']['company_addresses']['Row'][]
  company_ship_via: Database['public']['Tables']['company_ship_via']['Row'][]
}

interface CompaniesListProps {
  initialCompanies: ExternalCompany[]
}

export default function CompaniesList({ initialCompanies }: CompaniesListProps) {
  const [companies, setCompanies] = useState<ExternalCompany[]>(initialCompanies)
  const [filteredCompanies, setFilteredCompanies] = useState<ExternalCompany[]>(initialCompanies)
  const [searchTerm, setSearchTerm] = useState('')
  const [dialogOpen, setDialogOpen] = useState(false)
  const [editingCompany, setEditingCompany] = useState<ExternalCompany | null>(null)
  const [loading, setLoading] = useState(false)
  const [deleteLoading, setDeleteLoading] = useState<string | null>(null)
  const [error, setError] = useState<string | null>(null)

  useEffect(() => {
    const lowercasedTerm = searchTerm.toLowerCase()
    const filtered = companies.filter(company =>
      company.company_name.toLowerCase().includes(lowercasedTerm) ||
      (company.company_code && company.company_code.toLowerCase().includes(lowercasedTerm)) ||
      company.company_addresses.some(a => 
        (a.city && a.city.toLowerCase().includes(lowercasedTerm)) ||
        (a.country && a.country.toLowerCase().includes(lowercasedTerm))
      ) ||
      company.company_contacts.some(c => c.contact_name && c.contact_name.toLowerCase().includes(lowercasedTerm))
    )
    setFilteredCompanies(filtered)
  }, [companies, searchTerm])

  const fetchCompanies = async () => {
    try {
      setLoading(true)
      setError(null)
      
      // Fetch companies first
      const { data: companiesData, error: companiesError } = await supabase
        .from('companies')
        .select('*')
        .order('company_name')

      if (companiesError) {
        console.error('Fetch companies error:', companiesError)
        throw new Error(companiesError.message || 'Failed to fetch companies')
      }

      // Fetch addresses separately
      const { data: addressData } = await supabase
        .from('company_addresses')
        .select('*')
        .eq('company_ref_type', 'companies')

      // Fetch contacts separately 
      const { data: contactData } = await supabase
        .from('company_contacts') 
        .select('*')
        .eq('company_ref_type', 'companies')

      // Fetch shipping data separately
      const { data: shipViaData } = await supabase
        .from('company_ship_via')
        .select('*')
        .eq('company_ref_type', 'companies')

      // Combine the data
      const companiesWithRelations = companiesData?.map(company => ({
        ...company,
        company_addresses: addressData?.filter(addr => addr.company_id === company.company_id) || [],
        company_contacts: contactData?.filter(contact => contact.company_id === company.company_id) || [],
        company_ship_via: shipViaData?.filter(ship => ship.company_id === company.company_id) || []
      })) || []

      setCompanies(companiesWithRelations)
    } catch (error: any) {
      console.error('Error fetching companies:', error)
      const errorMessage = error.message || 'Failed to fetch companies'
      setError(errorMessage)
      toast.error(errorMessage)
    } finally {
      setLoading(false)
    }
  }

  const handleEdit = (company: ExternalCompany) => {
    setEditingCompany(company)
    setDialogOpen(true)
  }

  const handleDelete = async (company: ExternalCompany) => {
    if (!confirm(`Are you sure you want to delete ${company.company_name}?`)) return

    try {
      setDeleteLoading(company.company_id)
      
      // Check for references in purchase_orders as vendor
      const { data: vendorPOs, error: vendorPOError } = await supabase
        .from('purchase_orders')
        .select('po_id')
        .eq('vendor_company_id', company.company_id)
        .limit(1)

      if (vendorPOError) {
        console.error('Vendor PO check error:', vendorPOError)
        throw new Error(vendorPOError.message || 'Failed to check vendor PO references')
      }
      
      if (vendorPOs && vendorPOs.length > 0) {
        toast.error('Cannot delete company: It is referenced as a vendor in existing purchase orders.')
        return
      }

      // Check for references in purchase_orders as ship_via
      const { data: shipViaPOs, error: shipViaPOError } = await supabase
        .from('purchase_orders')
        .select('po_id')
        .eq('ship_via_id', company.company_id)
        .limit(1)

      if (shipViaPOError) {
        console.error('Ship via PO check error:', shipViaPOError)
        throw new Error(shipViaPOError.message || 'Failed to check ship via PO references')
      }
      
      if (shipViaPOs && shipViaPOs.length > 0) {
        toast.error('Cannot delete company: It is referenced as a shipping company in existing purchase orders.')
        return
      }

      // If no references found, proceed with deletion
      const { error } = await supabase
        .from('companies')
        .delete()
        .eq('company_id', company.company_id)

      if (error) {
        console.error('Delete company error:', error)
        throw new Error(error.message || 'Failed to delete company')
      }
      
      setCompanies(companies.filter(c => c.company_id !== company.company_id))
      toast.success('Company deleted successfully')
    } catch (error: any) {
      console.error('Error deleting company:', error)
      toast.error(error.message || 'Failed to delete company')
    } finally {
      setDeleteLoading(null)
    }
  }

  const handleDialogClose = () => {
    setDialogOpen(false)
    setEditingCompany(null)
    fetchCompanies()
  }

  return (
    <Card>
      <CardHeader>
        <div className="flex items-center justify-between">
          <div>
            <CardTitle className="text-base">Company List</CardTitle>
            <CardDescription>
              {companies.length} external companies â€¢ {filteredCompanies.length} shown
            </CardDescription>
          </div>
          <Button 
            onClick={() => {
              setEditingCompany(null)
              setDialogOpen(true)
            }}
            className="flex items-center gap-2"
          >
            <Plus className="h-4 w-4" />
            Add Company
          </Button>
        </div>
        <div className="relative">
          <Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-slate-400 h-4 w-4" />
          <Input placeholder="Search companies..." value={searchTerm} onChange={(e) => setSearchTerm(e.target.value)} className="pl-10" />
        </div>
      </CardHeader>
      <CardContent>
        {loading ? (
          <div className="text-center py-8">
            <div className="inline-block h-8 w-8 animate-spin border-4 border-current border-t-transparent rounded-full" />
            <div className="mt-2 text-slate-500">Loading companies...</div>
          </div>
        ) : error ? (
          <div className="text-center py-8">
            <div className="text-red-600 mb-2">Error loading companies</div>
            <div className="text-sm text-slate-500 mb-4">{error}</div>
            <Button variant="outline" onClick={fetchCompanies}>
              Try Again
            </Button>
          </div>
        ) : filteredCompanies.length === 0 ? (
          <div className="text-center py-8 text-slate-500">No companies found</div>
        ) : (
          <div className="space-y-3">
            {filteredCompanies.map((company) => (
              <Card key={company.company_id} className="hover:shadow-md transition-shadow">
                <CardHeader className="pb-2">
                  <div className="flex items-center justify-between">
                    <CardTitle className="text-base">{company.company_name}</CardTitle>
                    <div className="flex space-x-1">
                      <Button variant="ghost" size="icon" onClick={() => handleEdit(company)} disabled={deleteLoading === company.company_id}>
                        <Edit className="h-4 w-4" />
                      </Button>
                      <Button variant="ghost" size="icon" onClick={() => handleDelete(company)} disabled={deleteLoading === company.company_id}>
                        {deleteLoading === company.company_id ? (
                          <div className="h-4 w-4 animate-spin border-2 border-current border-t-transparent rounded-full" />
                        ) : (
                          <Trash2 className="h-4 w-4" />
                        )}
                      </Button>
                    </div>
                  </div>
                  <CardDescription className="text-xs">Code: {company.company_code} | Type: {company.company_type}</CardDescription>
                </CardHeader>
                <CardContent className="pt-2 text-xs text-slate-600 grid grid-cols-1 md:grid-cols-3 gap-4">
                  <div>
                    <h4 className="font-semibold mb-1">Addresses</h4>
                    {company.company_addresses.length > 0 ? (
                      <ul className="space-y-1">
                        {company.company_addresses.map(addr => (
                          <li key={addr.address_id}>{addr.address_line1}, {addr.city}, {addr.country}</li>
                        ))}
                      </ul>
                    ) : <p>No addresses</p>}
                  </div>
                  <div>
                    <h4 className="font-semibold mb-1">Contacts</h4>
                    {company.company_contacts.length > 0 ? (
                      <ul className="space-y-1">
                        {company.company_contacts.map(contact => (
                          <li key={contact.contact_id}>{contact.contact_name} ({contact.email})</li>
                        ))}
                      </ul>
                    ) : <p>No contacts</p>}
                  </div>
                  <div>
                    <h4 className="font-semibold mb-1">Shipping Methods</h4>
                    {company.company_ship_via.length > 0 ? (
                      <ul className="space-y-1">
                        {company.company_ship_via.map(ship => (
                          <li key={ship.ship_via_id} className="flex items-center gap-1">
                            <span className="text-blue-600 font-medium">{ship.ship_company_name}</span>
                            <span className="text-slate-500">({ship.account_no})</span>
                            {ship.ship_model && <span className="text-xs bg-slate-100 px-1 rounded">{ship.ship_model}</span>}
                          </li>
                        ))}
                      </ul>
                    ) : <p>No shipping methods</p>}
                  </div>
                </CardContent>
              </Card>
            ))}
          </div>
        )}
      </CardContent>
      <CompanyDialog open={dialogOpen} onClose={handleDialogClose} company={editingCompany as any} type="external_company" />
    </Card>
  )
}