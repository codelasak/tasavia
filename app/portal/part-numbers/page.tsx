import { createSupabaseServer } from '@/lib/supabase/server'
import { Database } from '@/lib/supabase/database.types'
import PartNumbersList from './part-numbers-list'

export const dynamic = 'force-dynamic'

type PartNumber = Database['public']['Tables']['pn_master_table']['Row']

async function getPartNumbers() {
  const supabase = createSupabaseServer()
  const { data, error } = await supabase
    .from('pn_master_table')
    .select('*')
    .order('pn')

  if (error) {
    console.error('Error fetching part numbers:', error)
    return []
  }
  return (data as any) || []
}

export default async function PartNumbersPage() {
  const partNumbers = await getPartNumbers()

  return (
    <div className="space-y-6">
      <div className="flex flex-col sm:flex-row sm:items-center sm:justify-between gap-2 pt-4 pb-4">
        <h1 className="text-2xl font-bold text-slate-900">Part Numbers</h1>
      </div>
      <PartNumbersList initialPartNumbers={partNumbers} />
    </div>
  )
}