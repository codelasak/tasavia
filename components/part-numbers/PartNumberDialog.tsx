'use client'

import { useEffect } from 'react'
import { useForm } from 'react-hook-form'
import { zodResolver } from '@hookform/resolvers/zod'
import * as z from 'zod'
import { Dialog, DialogContent, DialogDescription, DialogFooter, DialogHeader, DialogTitle } from '@/components/ui/dialog'
import { Button } from '@/components/ui/button'
import { Input } from '@/components/ui/input'
import { Label } from '@/components/ui/label'
import { Textarea } from '@/components/ui/textarea'
import { supabase } from '@/lib/supabase/client'
import { Database } from '@/lib/supabase/database.types'
import { toast } from 'sonner'

type PartNumber = Database['public']['Tables']['pn_master_table']['Row']

const partNumberSchema = z.object({
  pn: z.string().min(1, 'Part number is required'),
  description: z.string().optional(),
  remarks: z.string().optional(),
})

type PartNumberFormValues = z.infer<typeof partNumberSchema>

interface PartNumberDialogProps {
  open: boolean
  onClose: () => void
  partNumber?: PartNumber | null
}

export function PartNumberDialog({ open, onClose, partNumber }: PartNumberDialogProps) {
  const form = useForm<PartNumberFormValues>({
    resolver: zodResolver(partNumberSchema),
    defaultValues: {
      pn: '',
      description: '',
      remarks: '',
    }
  })

  useEffect(() => {
    if (partNumber) {
      form.reset({
        pn: partNumber.pn,
        description: partNumber.description || '',
        remarks: partNumber.remarks || '',
      })
    } else {
      form.reset({
        pn: '',
        description: '',
        remarks: '',
      })
    }
  }, [partNumber, form])

  const onSubmit = async (data: PartNumberFormValues) => {
    try {
      const submitData = {
        ...data,
        description: data.description || null,
        remarks: data.remarks || null,
      }

      if (partNumber) {
        const { error } = await supabase
          .from('pn_master_table')
          .update(submitData)
          .eq('pn_id', partNumber.pn_id)
        if (error) throw error
        toast.success('Part number updated successfully')
      } else {
        const { error } = await supabase
          .from('pn_master_table')
          .insert(submitData)
        if (error) throw error
        toast.success('Part number created successfully')
      }
      onClose()
    } catch (error: any) {
      console.error('Error saving part number:', error)
      if (error.code === '23505') {
        toast.error('This part number already exists')
      } else {
        toast.error(error.message || 'Failed to save part number')
      }
    }
  }

  return (
    <Dialog open={open} onOpenChange={onClose}>
      <DialogContent className="sm:max-w-[600px]">
        <DialogHeader>
          <DialogTitle>
            {partNumber ? 'Edit' : 'Add'} Part Number
          </DialogTitle>
          <DialogDescription>
            {partNumber ? 'Update' : 'Create'} part number information.
          </DialogDescription>
        </DialogHeader>
        
        <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-4">
          <div>
            <Label htmlFor="pn">Part Number</Label>
            <Input
              id="pn"
              {...form.register('pn')}
              className={form.formState.errors.pn ? 'border-red-500' : ''}
              placeholder="Enter part number"
            />
            {form.formState.errors.pn && (
              <div className="text-red-500 text-sm mt-1">
                {form.formState.errors.pn.message}
              </div>
            )}
          </div>

          <div>
            <Label htmlFor="description">Description</Label>
            <Textarea
              id="description"
              {...form.register('description')}
              rows={3}
              placeholder="Enter part description"
            />
          </div>

          <div>
            <Label htmlFor="remarks">Remarks</Label>
            <Textarea
              id="remarks"
              {...form.register('remarks')}
              rows={3}
              placeholder="Enter any additional remarks"
            />
          </div>

          <DialogFooter>
            <Button type="button" variant="outline" onClick={onClose}>
              Cancel
            </Button>
            <Button type="submit" disabled={form.formState.isSubmitting}>
              {form.formState.isSubmitting ? 'Saving...' : partNumber ? 'Update' : 'Create'}
            </Button>
          </DialogFooter>
        </form>
      </DialogContent>
    </Dialog>
  )
}