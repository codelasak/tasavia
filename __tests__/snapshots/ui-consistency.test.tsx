/**
 * Snapshot tests for UI consistency
 * These tests ensure that component UIs remain consistent across changes
 */

import React from 'react'
import { render, act } from '@testing-library/react'
import { CompanyDialog } from '../../components/companies/CompanyDialog'
import { PartNumberDialog } from '../../components/part-numbers/PartNumberDialog'
import CompaniesList from '../../app/portal/companies/companies-list'
import MyCompaniesList from '../../app/portal/my-companies/my-companies-list'
import PartNumbersList from '../../app/portal/part-numbers/part-numbers-list'

// Mock dependencies
jest.mock('@/lib/supabase/client', () => ({
  supabase: {
    from: jest.fn().mockReturnValue({
      select: jest.fn().mockReturnValue({
        eq: jest.fn().mockReturnValue({
          eq: jest.fn().mockResolvedValue({ data: [], error: null }),
        }),
        order: jest.fn().mockResolvedValue({ data: [], error: null }),
      }),
      insert: jest.fn().mockResolvedValue({ error: null }),
      update: jest.fn().mockReturnValue({
        eq: jest.fn().mockResolvedValue({ error: null }),
      }),
      delete: jest.fn().mockReturnValue({
        eq: jest.fn().mockResolvedValue({ error: null }),
      }),
    }),
  },
}))

jest.mock('sonner', () => ({
  toast: {
    success: jest.fn(),
    error: jest.fn(),
  },
}))

// Sample data for consistent testing
const sampleExternalCompany = {
  company_id: '1',
  company_name: 'Acme Corporation',
  company_code: 'ACM001',
  company_type: 'vendor' as const,
  created_at: '2024-01-01T00:00:00Z',
  updated_at: '2024-01-01T00:00:00Z',
  company_contacts: [
    {
      contact_id: 'c1',
      contact_name: 'John Doe',
      email: 'john@acme.com',
      phone: '+1234567890',
      role: 'Manager',
      is_primary: true,
      company_id: '1',
      company_ref_type: 'companies' as const,
      created_at: '2024-01-01T00:00:00Z',
      updated_at: '2024-01-01T00:00:00Z',
    },
  ],
  company_addresses: [
    {
      address_id: 'a1',
      address_line1: '123 Main St',
      address_line2: 'Suite 100',
      city: 'New York',
      country: 'USA',
      zip_code: '10001',
      is_primary: true,
      company_id: '1',
      company_ref_type: 'companies' as const,
      created_at: '2024-01-01T00:00:00Z',
      updated_at: '2024-01-01T00:00:00Z',
    },
  ],
  company_ship_via: [
    {
      ship_via_id: 's1',
      ship_company_name: 'DHL',
      predefined_company: 'DHL' as const,
      custom_company_name: null,
      account_no: '123456789',
      owner: 'Acme Corp',
      ship_model: 'GROUND' as const,
      company_id: '1',
      company_ref_type: 'companies' as const,
      created_at: '2024-01-01T00:00:00Z',
      updated_at: '2024-01-01T00:00:00Z',
    },
  ],
}

const sampleMyCompany = {
  my_company_id: '1',
  my_company_name: 'My Test Company',
  my_company_code: 'MTC001',
  created_at: '2024-01-01T00:00:00Z',
  updated_at: '2024-01-01T00:00:00Z',
  company_contacts: [
    {
      contact_id: 'c1',
      contact_name: 'Jane Smith',
      email: 'jane@mytestcompany.com',
      phone: '+1987654321',
      role: 'CEO',
      is_primary: true,
      company_id: '1',
      company_ref_type: 'my_companies' as const,
      created_at: '2024-01-01T00:00:00Z',
      updated_at: '2024-01-01T00:00:00Z',
    },
  ],
  company_addresses: [
    {
      address_id: 'a1',
      address_line1: '456 Business Ave',
      address_line2: 'Floor 5',
      city: 'San Francisco',
      country: 'USA',
      zip_code: '94105',
      is_primary: true,
      company_id: '1',
      company_ref_type: 'my_companies' as const,
      created_at: '2024-01-01T00:00:00Z',
      updated_at: '2024-01-01T00:00:00Z',
    },
  ],
  company_ship_via: [
    {
      ship_via_id: 's1',
      ship_company_name: 'FedEx',
      predefined_company: 'FEDEX' as const,
      custom_company_name: null,
      account_no: '987654321',
      owner: 'My Test Company',
      ship_model: 'GROUND' as const,
      company_id: '1',
      company_ref_type: 'my_companies' as const,
      created_at: '2024-01-01T00:00:00Z',
      updated_at: '2024-01-01T00:00:00Z',
    },
  ],
}

const samplePartNumbers = [
  {
    pn_id: '1',
    pn: 'PN-001',
    description: 'High-quality bearing assembly',
    remarks: 'Compatible with models X100-X200',
    created_at: '2024-01-01T00:00:00Z',
    updated_at: '2024-01-01T00:00:00Z',
  },
  {
    pn_id: '2',
    pn: 'GASKET-301',
    description: 'Rubber gasket seal',
    remarks: null,
    created_at: '2024-01-02T00:00:00Z',
    updated_at: '2024-01-02T00:00:00Z',
  },
]

describe('UI Consistency Snapshots', () => {
  describe('Company Dialog Snapshots', () => {
    it('should match snapshot for new external company dialog', () => {
      let component
      act(() => {
        component = render(
          <CompanyDialog
            open={true}
            onClose={() => {}}
            company={null}
            type="external_company"
          />
        )
      })
      expect(component.container.firstChild).toMatchSnapshot('external-company-dialog-new')
    })

    it('should match snapshot for editing external company dialog', () => {
      let component
      act(() => {
        component = render(
          <CompanyDialog
            open={true}
            onClose={() => {}}
            company={sampleExternalCompany}
            type="external_company"
          />
        )
      })
      expect(component.container.firstChild).toMatchSnapshot('external-company-dialog-edit')
    })

    it('should match snapshot for new my company dialog', () => {
      let component
      act(() => {
        component = render(
          <CompanyDialog
            open={true}
            onClose={() => {}}
            company={null}
            type="my_company"
          />
        )
      })
      expect(component.container.firstChild).toMatchSnapshot('my-company-dialog-new')
    })

    it('should match snapshot for editing my company dialog', () => {
      let component
      act(() => {
        component = render(
          <CompanyDialog
            open={true}
            onClose={() => {}}
            company={sampleMyCompany}
            type="my_company"
          />
        )
      })
      expect(component.container.firstChild).toMatchSnapshot('my-company-dialog-edit')
    })

    it('should match snapshot for closed dialog', () => {
      let component
      act(() => {
        component = render(
          <CompanyDialog
            open={false}
            onClose={() => {}}
            company={null}
            type="external_company"
          />
        )
      })
      expect(component.container.firstChild).toMatchSnapshot('company-dialog-closed')
    })
  })

  describe('Part Number Dialog Snapshots', () => {
    it('should match snapshot for new part number dialog', () => {
      const { container } = render(
        <PartNumberDialog
          open={true}
          onClose={() => {}}
          partNumber={null}
        />
      )
      expect(container.firstChild).toMatchSnapshot('part-number-dialog-new')
    })

    it('should match snapshot for editing part number dialog', () => {
      const { container } = render(
        <PartNumberDialog
          open={true}
          onClose={() => {}}
          partNumber={samplePartNumbers[0]}
        />
      )
      expect(container.firstChild).toMatchSnapshot('part-number-dialog-edit')
    })

    it('should match snapshot for closed part number dialog', () => {
      const { container } = render(
        <PartNumberDialog
          open={false}
          onClose={() => {}}
          partNumber={null}
        />
      )
      expect(container.firstChild).toMatchSnapshot('part-number-dialog-closed')
    })
  })

  describe('Companies List Snapshots', () => {
    it('should match snapshot for external companies list with data', () => {
      const { container } = render(
        <CompaniesList initialCompanies={[sampleExternalCompany]} />
      )
      expect(container.firstChild).toMatchSnapshot('external-companies-list-with-data')
    })

    it('should match snapshot for external companies list empty', () => {
      const { container } = render(
        <CompaniesList initialCompanies={[]} />
      )
      expect(container.firstChild).toMatchSnapshot('external-companies-list-empty')
    })

    it('should match snapshot for my companies list with data', () => {
      const { container } = render(
        <MyCompaniesList initialCompanies={[sampleMyCompany]} />
      )
      expect(container.firstChild).toMatchSnapshot('my-companies-list-with-data')
    })

    it('should match snapshot for my companies list empty', () => {
      const { container } = render(
        <MyCompaniesList initialCompanies={[]} />
      )
      expect(container.firstChild).toMatchSnapshot('my-companies-list-empty')
    })
  })

  describe('Part Numbers List Snapshots', () => {
    it('should match snapshot for part numbers list with data', () => {
      const { container } = render(
        <PartNumbersList initialPartNumbers={samplePartNumbers} />
      )
      expect(container.firstChild).toMatchSnapshot('part-numbers-list-with-data')
    })

    it('should match snapshot for part numbers list empty', () => {
      const { container } = render(
        <PartNumbersList initialPartNumbers={[]} />
      )
      expect(container.firstChild).toMatchSnapshot('part-numbers-list-empty')
    })

    it('should match snapshot for part numbers with long descriptions', () => {
      const longDescriptionPartNumbers = [
        {
          ...samplePartNumbers[0],
          description: 'This is a very long description that should test how the component handles text wrapping and truncation in the UI layout',
          remarks: 'This is also a very long remark that should test the text handling capabilities of the component when dealing with extensive textual content',
        },
      ]

      const { container } = render(
        <PartNumbersList initialPartNumbers={longDescriptionPartNumbers} />
      )
      expect(container.firstChild).toMatchSnapshot('part-numbers-list-long-text')
    })
  })

  describe('Component States Snapshots', () => {
    it('should match snapshot for companies list with multiple shipping methods', () => {
      const companyWithMultipleShipping = {
        ...sampleExternalCompany,
        company_ship_via: [
          ...sampleExternalCompany.company_ship_via,
          {
            ship_via_id: 's2',
            ship_company_name: 'Custom Courier',
            predefined_company: 'CUSTOM' as const,
            custom_company_name: 'Custom Courier',
            account_no: 'CC-12345',
            owner: 'Acme Corp',
            ship_model: 'SEA' as const,
            company_id: '1',
            company_ref_type: 'companies' as const,
            created_at: '2024-01-01T00:00:00Z',
            updated_at: '2024-01-01T00:00:00Z',
          },
        ],
      }

      const { container } = render(
        <CompaniesList initialCompanies={[companyWithMultipleShipping]} />
      )
      expect(container.firstChild).toMatchSnapshot('companies-list-multiple-shipping')
    })

    it('should match snapshot for company with no contacts or addresses', () => {
      const minimalCompany = {
        ...sampleExternalCompany,
        company_contacts: [],
        company_addresses: [],
        company_ship_via: [],
      }

      const { container } = render(
        <CompaniesList initialCompanies={[minimalCompany]} />
      )
      expect(container.firstChild).toMatchSnapshot('companies-list-minimal-data')
    })

    it('should match snapshot for mixed company types', () => {
      const mixedCompanies = [
        sampleExternalCompany,
        {
          ...sampleExternalCompany,
          company_id: '2',
          company_name: 'Beta Industries',
          company_type: 'customer' as const,
          company_contacts: [],
          company_addresses: [],
          company_ship_via: [],
        },
        {
          ...sampleExternalCompany,
          company_id: '3',
          company_name: 'Gamma Corp',
          company_type: 'both' as const,
        },
      ]

      const { container } = render(
        <CompaniesList initialCompanies={mixedCompanies} />
      )
      expect(container.firstChild).toMatchSnapshot('companies-list-mixed-types')
    })
  })

  describe('Responsive Layout Snapshots', () => {
    it('should match snapshot for wide screen layout', () => {
      // Mock window width for testing
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1920,
      })

      const { container } = render(
        <CompaniesList initialCompanies={[sampleExternalCompany]} />
      )
      expect(container.firstChild).toMatchSnapshot('companies-list-wide-screen')
    })

    it('should match snapshot for mobile layout', () => {
      // Mock window width for mobile
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 375,
      })

      const { container } = render(
        <PartNumbersList initialPartNumbers={samplePartNumbers} />
      )
      expect(container.firstChild).toMatchSnapshot('part-numbers-list-mobile')
    })
  })

  describe('Edge Cases Snapshots', () => {
    it('should match snapshot for very long company names', () => {
      const longNameCompany = {
        ...sampleExternalCompany,
        company_name: 'This Is A Very Long Company Name That Should Test Text Wrapping And Layout Handling',
        company_code: 'VERY-LONG-CODE-123456789',
      }

      const { container } = render(
        <CompaniesList initialCompanies={[longNameCompany]} />
      )
      expect(container.firstChild).toMatchSnapshot('companies-list-long-names')
    })

    it('should match snapshot for special characters in part numbers', () => {
      const specialCharPartNumbers = [
        {
          pn_id: '1',
          pn: 'PN-001/A-B_C.D',
          description: 'Part with special chars: @#$%^&*()',
          remarks: 'Remarks with émojis 🔧⚙️ and ñ special chars',
          created_at: '2024-01-01T00:00:00Z',
          updated_at: '2024-01-01T00:00:00Z',
        },
      ]

      const { container } = render(
        <PartNumbersList initialPartNumbers={specialCharPartNumbers} />
      )
      expect(container.firstChild).toMatchSnapshot('part-numbers-list-special-chars')
    })
  })
})

// Custom snapshot serializer for consistent formatting
expect.addSnapshotSerializer({
  test: (val) => val && val.hasAttribute && val.hasAttribute('class'),
  print: (val) => {
    const element = val as Element
    const tagName = element.tagName.toLowerCase()
    const className = element.getAttribute('class') || ''
    const textContent = element.textContent?.slice(0, 50) || ''
    
    return `<${tagName}${className ? ` class="${className}"` : ''}>${textContent}${textContent.length > 50 ? '...' : ''}</${tagName}>`
  },
})