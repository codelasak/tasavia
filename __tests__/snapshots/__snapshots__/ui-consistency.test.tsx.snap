// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UI Consistency Snapshots Companies List Snapshots should match snapshot for external companies list empty: external-companies-list-empty 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List0 external companies • 0 shownAdd Comp</div>`;

exports[`UI Consistency Snapshots Companies List Snapshots should match snapshot for external companies list with data: external-companies-list-with-data 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List1 external companies • 1 shownAdd Comp</div>`;

exports[`UI Consistency Snapshots Companies List Snapshots should match snapshot for my companies list empty: my-companies-list-empty 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List0 my companies • 0 shownAdd CompanyNo </div>`;

exports[`UI Consistency Snapshots Companies List Snapshots should match snapshot for my companies list with data: my-companies-list-with-data 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List1 my companies • 1 shownAdd CompanyMy </div>`;

exports[`UI Consistency Snapshots Company Dialog Snapshots should match snapshot for closed dialog: company-dialog-closed 1`] = `null`;

exports[`UI Consistency Snapshots Company Dialog Snapshots should match snapshot for editing external company dialog: external-company-dialog-edit 1`] = `null`;

exports[`UI Consistency Snapshots Company Dialog Snapshots should match snapshot for editing my company dialog: my-company-dialog-edit 1`] = `null`;

exports[`UI Consistency Snapshots Company Dialog Snapshots should match snapshot for new external company dialog: external-company-dialog-new 1`] = `null`;

exports[`UI Consistency Snapshots Company Dialog Snapshots should match snapshot for new my company dialog: my-company-dialog-new 1`] = `null`;

exports[`UI Consistency Snapshots Component States Snapshots should match snapshot for companies list with multiple shipping methods: companies-list-multiple-shipping 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List1 external companies • 1 shownAdd Comp</div>`;

exports[`UI Consistency Snapshots Component States Snapshots should match snapshot for company with no contacts or addresses: companies-list-minimal-data 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List1 external companies • 1 shownAdd Comp</div>`;

exports[`UI Consistency Snapshots Component States Snapshots should match snapshot for mixed company types: companies-list-mixed-types 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List3 external companies • 3 shownAdd Comp</div>`;

exports[`UI Consistency Snapshots Edge Cases Snapshots should match snapshot for special characters in part numbers: part-numbers-list-special-chars 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Part Numbers1 part numbers • 1 shownAdd Part Numbe</div>`;

exports[`UI Consistency Snapshots Edge Cases Snapshots should match snapshot for very long company names: companies-list-long-names 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List1 external companies • 1 shownAdd Comp</div>`;

exports[`UI Consistency Snapshots Part Number Dialog Snapshots should match snapshot for closed part number dialog: part-number-dialog-closed 1`] = `null`;

exports[`UI Consistency Snapshots Part Number Dialog Snapshots should match snapshot for editing part number dialog: part-number-dialog-edit 1`] = `null`;

exports[`UI Consistency Snapshots Part Number Dialog Snapshots should match snapshot for new part number dialog: part-number-dialog-new 1`] = `null`;

exports[`UI Consistency Snapshots Part Numbers List Snapshots should match snapshot for part numbers list empty: part-numbers-list-empty 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Part Numbers0 part numbers • 0 shownAdd Part Numbe</div>`;

exports[`UI Consistency Snapshots Part Numbers List Snapshots should match snapshot for part numbers list with data: part-numbers-list-with-data 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Part Numbers2 part numbers • 2 shownAdd Part Numbe</div>`;

exports[`UI Consistency Snapshots Part Numbers List Snapshots should match snapshot for part numbers with long descriptions: part-numbers-list-long-text 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Part Numbers1 part numbers • 1 shownAdd Part Numbe</div>`;

exports[`UI Consistency Snapshots Responsive Layout Snapshots should match snapshot for mobile layout: part-numbers-list-mobile 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Part Numbers2 part numbers • 2 shownAdd Part Numbe</div>`;

exports[`UI Consistency Snapshots Responsive Layout Snapshots should match snapshot for wide screen layout: companies-list-wide-screen 1`] = `<div class="rounded-lg border bg-card text-card-foreground shadow-sm">Company List1 external companies • 1 shownAdd Comp</div>`;
